scipy/stats/_qmc.py

```diff
@@ -1036,6 +1084,56 @@ def _random(self, n: IntNumber = 1) -> np.ndarray:
         self.num_generated += n
         return samples
 
+    def _random_oa_lhs(self, n: IntNumber = 1) -> np.ndarray:
+        """Orthogonal array based LHS of strength 2."""
+        p = np.sqrt(n).astype(int)
+        n_row = p**2
+        n_col = p + 1
+
+        primes = primes_from_2_to(p + 2)
+        if p not in primes or n != n_row:
+            raise ValueError(
+                "n is not the square of a prime number. Closest"
+                f" values are {primes[-2:]**2}"
+            )
+        if self.d > p + 1:
+            raise ValueError(
+                "d is too large. Must be <= sqrt(n) + 1"
+            )
+
+        oa_sample = np.zeros(shape=(n_row, n_col), dtype=int)
+
+        # OA of strength 2
+        arrays = np.tile(np.arange(p), (2, 1))
+        oa_sample[:, :2] = np.stack(np.meshgrid(*arrays),
+                                    axis=-1).reshape(-1, 2)
+        for p_ in range(1, p):
+            oa_sample[:, 2+p_-1] = np.mod(oa_sample[:, 0]
+                                          + p_*oa_sample[:, 1], p)
+
+        # scramble the OA
+        oa_sample_ = np.empty(shape=(n_row, n_col), dtype=int)
+        for j in range(n_col):
+            perms = self.rng.permutation(p)
+            oa_sample_[:, j] = perms[oa_sample[:, j]]
+
+        # following is making a scrambled OA into an OA-LHS
+        oa_lhs_sample = np.zeros(shape=(n_row, n_col))
+        lhs_engine = LatinHypercube(d=1, centered=self.centered,
+                                    orthogonal=False,
+                                    seed=self.rng)  # type: QMCEngine
+        for j in range(n_col):
+            for k in range(p):
+                idx = np.where(oa_sample[:, j] == k)[0]
+                lhs = lhs_engine.random(p).flatten()
+                oa_lhs_sample[:, j][idx] = lhs + oa_sample[:, j][idx]
```

Comment:
I do not think you need to use LHS for the noise term here. Is there a reason for this? I feel that just using random numbers could simplify the code.

Reply:
This is from Art Owen. Here is an extract from our communication

> Let A be an orthogonal array of strength 2 with n = p^2 rows and p+1 columns from the Bose construction with prime p.
Elements of A are integers 0,1,2,...,p-1
Each column of A has been permuted into random order
>
> Now:
> 
> For each column j = 1,...,p+1  of A
    ..for each integer level k=0,1,...p-1
    ....find the p rows of A where column j has value k
    ....get p stratified U(0,1)'s
    ....add them in random order to the elements of A[ : j ]  in those p rows
>
> Then divide A by p.

And also
> Or ... for each column generate an LHS of size n=p^2 and d=1.  Now replace the k'th largest randomized orthogonal array value in the column by the k'th largest LHS value.  Do that for each column and you'll have one of Tang's OA based LHS.

Reply:
Thank you for the clarification! These lines make sense now.